core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -562,6 +564,10 @@ class ReplicaManager(val config: KafkaConfig,
    * Append messages to leader replicas of the partition, and wait for them to be replicated to other replicas;
    * the callback function will be triggered either when timeout or the required acks are satisfied;
    * if the callback function itself is already synchronized on some object then pass this object to avoid deadlock.
+   *
+   * Noted that all pending delayed check operations are stored in a queue. All callers to ReplicaManager.appendRecords()
+   * are expected to take up to 1 item from that queue and check the completeness for all affected partitions, without
```

Comment:
What's the reasoning for taking just 1 item @junrao? Could this cause the queue to grow over time?

Reply:
Good question. It's based on the assumption that each KafkaApis.handle() call only calls ReplicaManager. appendRecords() once. Not sure if this is always true in the future. Perhaps a safer approach is to have Action.tryCompleteAction() get the current size of the queue and complete all those actions.

Reply:
It should be fine to let handler complete actions as much as possible since the response is created before handling delayed actions.